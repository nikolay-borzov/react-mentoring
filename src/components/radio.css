.radio-input {
  margin-right: var(--margin-input);
}

.radio-input input {
  opacity: 0;
  position: absolute;
}

.radio-input__label {
  user-select: none;
}

.radio-input--style-button .radio-input__label {
  display: flex;
  padding: var(--padding-button-small);
  text-transform: uppercase;
  font-size: 1em;
  font-weight: 700;
  transition: background-color var(--transition-duration),
    box-shadow var(--transition-duration);
  box-sizing: border-box;
  border: none;
  color: var(--color-text-light);
  background-color: var(--color-default);
}

.radio-input--style-button .radio-input__label:hover {
  box-shadow: 0 0 20px var(--color-primary), 2px 2px 2px #000;
}

.radio-input--style-button input:checked + .radio-input__label {
  background-color: var(--color-primary);
}

.radio-input--style-plain .radio-input__label {
  padding: 0.25em 0.5em;
}

.radio-input--style-plain input:checked + .radio-input__label {
  color: var(--color-primary);
}

.radio-input input:focus + .radio-input__label {
  outline: var(--focus);
}
