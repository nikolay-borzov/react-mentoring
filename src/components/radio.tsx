import * as React from 'react'
import styled, { css, FlattenSimpleInterpolation } from 'styled-components'

import { HoverEffectMixin, FormLabel } from '../styles'

const STYLE_BUTTON = 'button'
const STYLE_PLAIN = 'plain'

interface RadioOption {
  name: string
  value: string
}

type RadioStyle = 'button' | 'plain'

interface RadioPropsCommon {
  label: string
  name: string
  options: RadioOption[]
  style?: RadioStyle
}

export interface RadioControlledProps extends RadioPropsCommon {
  label: string
  name: string
  value?: string
  options: RadioOption[]
  style?: RadioStyle
  onChange(value: string): void
}

export interface RadioUncontrolledProps extends RadioPropsCommon {
  label: string
  name: string
  defaultValue?: string
  options: RadioOption[]
  style?: RadioStyle
}

export type RadioProps = RadioControlledProps | RadioUncontrolledProps

const RadioInputWrapper = styled.div`
  display: flex;
  align-items: center;
`

const labelMixinMap: Record<string, FlattenSimpleInterpolation> = {
  [STYLE_BUTTON]: css`
    display: flex;
    padding: var(--padding-button-small);
    text-transform: uppercase;
    font-size: 1rem;
    font-weight: 700;
    box-sizing: border-box;
    border: none;
    color: var(--color-text-light);
    background-color: var(--color-default);

    ${HoverEffectMixin};

    input:checked + & {
      background-color: var(--color-primary);
    }
  `,
  [STYLE_PLAIN]: css`
    padding: 0.25rem 0.5rem;

    input:checked + & {
      color: var(--color-primary);
    }
  `
}

interface RadioInputLabelProps {
  inputStyle: RadioStyle
}

const RadioInputLabel = styled.label<RadioInputLabelProps>`
  user-select: none;
  white-space: nowrap;

  ${props => labelMixinMap[props.inputStyle]};
`

const RadioInput = styled.div`
  margin-right: var(--margin-input);

  input {
    opacity: 0;
    position: absolute;
  }

  input:focus + ${RadioInputLabel} {
    outline: var(--focus);
  }
`

function isControlledRadio(props: RadioProps): props is RadioControlledProps {
  return 'onChange' in props
}

class RadioBase<TProps extends RadioPropsCommon> extends React.PureComponent<TProps> {
  value?: string

  isDefaultChecked(option: RadioOption) {
    return option.value === this.value
  }

  getOptionInputId(option: RadioOption) {
    return `${this.props.name}_${option.value}`
  }

  onChange = (changeEvent: React.ChangeEvent<HTMLInputElement>) => { }

  render() {
    const { label, options, style = 'plain' } = this.props

    return (
      <RadioInputWrapper>
        <FormLabel className="nowrap">{label}</FormLabel>
        {options.map(option => (
          <RadioInput key={option.value}>
            <input
              type="radio"
              id={this.getOptionInputId(option)}
              name={this.props.name}
              value={option.value}
              defaultChecked={this.isDefaultChecked(option)}
              onChange={this.onChange}
            />
            <RadioInputLabel
              htmlFor={this.getOptionInputId(option)}
              inputStyle={style}>
              {option.name}
            </RadioInputLabel>
          </RadioInput>
        ))}
      </RadioInputWrapper>
    )
  }
}

export class RadioControlled extends RadioBase<RadioControlledProps> {
  constructor(props: RadioControlledProps) {
    super(props)

    this.value = props.value
  }

  onChange = (changeEvent: React.ChangeEvent<HTMLInputElement>) => {
    const value = changeEvent.target.value

    this.props.onChange(value)
  }
}

export class RadioUncontrolled extends RadioBase<RadioUncontrolledProps> {
  constructor(props: RadioUncontrolledProps) {
    super(props)

    this.value = props.defaultValue
  }

  onChange = (changeEvent: React.ChangeEvent<HTMLInputElement>) => {
    const value = changeEvent.target.value

    this.value = value
  }
}

// TODO: Replace by two components
export class Radio extends React.PureComponent<RadioProps> {
  value?: string

  constructor(props: RadioProps) {
    super(props)

    let value
    if (isControlledRadio(props)) {
      value = props.value
    } else {
      value = props.defaultValue
    }

    // Set 'value' so that selected value is available in form's onSubmit handler
    this.value = value
  }

  isDefaultChecked(option: RadioOption) {
    return option.value === this.value
  }

  getOptionInputId(option: RadioOption) {
    return `${this.props.name}_${option.value}`
  }

  onChange = (changeEvent: React.ChangeEvent<HTMLInputElement>) => {
    const value = changeEvent.target.value

    if (isControlledRadio(this.props)) {
      this.props.onChange(value)
    } else {
      this.value = value
    }
  }

  render() {
    const { label, options, style = 'plain' } = this.props

    return (
      <RadioInputWrapper>
        <FormLabel className="nowrap">{label}</FormLabel>
        {options.map(option => (
          <RadioInput key={option.value}>
            <input
              type="radio"
              id={this.getOptionInputId(option)}
              name={this.props.name}
              value={option.value}
              defaultChecked={this.isDefaultChecked(option)}
              onChange={this.onChange}
            />
            <RadioInputLabel
              htmlFor={this.getOptionInputId(option)}
              inputStyle={style}>
              {option.name}
            </RadioInputLabel>
          </RadioInput>
        ))}
      </RadioInputWrapper>
    )
  }
}
