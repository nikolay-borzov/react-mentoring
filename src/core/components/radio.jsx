import React from 'react'
import PropTypes from 'prop-types'

const styleClassMap = {
  button: 'radio-input--style-button',
  plain: 'radio-input--style-plain'
}

export class Radio extends React.PureComponent {
  static propTypes = {
    label: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    options: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        value: PropTypes.string.isRequired,
        selected: PropTypes.bool
      })
    ).isRequired,
    style: PropTypes.oneOf(['button', 'plain'])
  }

  static defaultProps = {
    style: 'plain'
  }

  constructor(props) {
    super(props)

    this.styleClassName = styleClassMap[props.style]
  }

  isOptionSelected(option) {
    return option.value === this.props.value
  }

  render() {
    return (
      <React.Fragment>
        <label className="form-label">{this.props.label}</label>
        {this.props.options.map(option => (
          <div
            key={option.value}
            className={`radio-input ${this.styleClassName}`}
          >
            <input
              type="radio"
              id={`${this.props.name}_${option.value}`}
              name={this.props.name}
              value={option.value}
              defaultChecked={this.isOptionSelected(option)}
            />
            <label
              htmlFor={`${this.props.name}_${option.value}`}
              className="radio-input__label"
            >
              {option.name}
            </label>
          </div>
        ))}
      </React.Fragment>
    )
  }
}
