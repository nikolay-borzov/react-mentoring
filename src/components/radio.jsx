import React from 'react'
import PropTypes from 'prop-types'
import requiredIf from 'react-required-if'
import { isUndefined } from 'lodash-es'

import './radio.css'

const styleClassMap = {
  button: 'radio-input--style-button',
  plain: 'radio-input--style-plain'
}

export class Radio extends React.PureComponent {
  static propTypes = {
    label: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    value: requiredIf(PropTypes.string, props => !isUndefined(props.onChange)),
    defaultValue: requiredIf(
      PropTypes.string,
      props => isUndefined(props.onChange) && isUndefined(props.value)
    ),
    options: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        value: PropTypes.string.isRequired
      })
    ).isRequired,
    onChange: requiredIf(PropTypes.func, props => !isUndefined(props.value)),
    style: PropTypes.oneOf(['button', 'plain'])
  }

  static defaultProps = {
    style: 'plain'
  }

  constructor(props) {
    super(props)

    this.isControlled =
      !isUndefined(props.value) && !isUndefined(props.onChange)

    // Set 'value' so that selected value is available in form's onSubmit handler
    this.value = this.isControlled ? this.props.value : this.props.defaultValue
    this.styleClassName = styleClassMap[props.style]
  }

  isDefaultChecked(option) {
    return option.value === this.value
  }

  getOptionInputId(option) {
    return `${this.props.name}_${option.value}`
  }

  onChange = changeEvent => {
    const value = changeEvent.target.value

    if (this.isControlled) {
      this.props.onChange(value)
    } else {
      this.value = value
    }
  }

  render() {
    return (
      <div className="radio-input-wrapper">
        <label className="form-label nowrap">{this.props.label}</label>
        {this.props.options.map(option => (
          <div
            key={option.value}
            className={`radio-input ${this.styleClassName}`}>
            <input
              type="radio"
              id={this.getOptionInputId(option)}
              name={this.props.name}
              value={option.value}
              defaultChecked={this.isDefaultChecked(option)}
              onChange={this.onChange}
            />
            <label
              htmlFor={this.getOptionInputId(option)}
              className="radio-input__label">
              {option.name}
            </label>
          </div>
        ))}
      </div>
    )
  }
}
