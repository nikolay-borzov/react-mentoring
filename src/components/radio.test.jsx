import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import { Radio } from './radio'

describe('Radio component', () => {
  const options = [
    {
      name: 'Cat',
      value: 'cat'
    },
    {
      name: 'Dog',
      value: 'dog'
    }
  ]

  const selectLastRadio = (wrapper, value) => {
    wrapper
      .find('input[type="radio"]')
      .last()
      .simulate('change', { target: { value: 'dog' } })
  }

  let props

  beforeEach(() => {
    props = {
      label: 'Radio label',
      name: 'radioName',
      options: options,
      value: undefined,
      onChange: undefined,
      defaultValue: undefined
    }
  })

  describe('if controlled', () => {
    it('renders correctly', () => {
      props.value = 'cat'
      props.onChange = jest.fn()
      props.style = 'button'

      const tree = renderer.create(<Radio {...props} />).toJSON()

      expect(tree).toMatchSnapshot()
    })

    it(`calls 'onChange' callback`, () => {
      const onChangeMock = jest.fn()
      props.value = 'cat'
      props.onChange = onChangeMock
      const wrapper = shallow(<Radio {...props} />)

      selectLastRadio(wrapper)

      expect(onChangeMock.mock.calls.length).toBe(1)
      expect(onChangeMock.mock.calls[0][0]).toBe('dog')
    })
  })

  describe('if uncontrolled', () => {
    it('renders correctly', () => {
      props.defaultValue = 'cat'
      props.style = 'plain'

      const tree = renderer.create(<Radio {...props} />).toJSON()

      expect(tree).toMatchSnapshot()
    })

    it(`sets 'value' to be equal to 'defaultValue'`, () => {
      props.defaultValue = 'cat'
      const wrapper = shallow(<Radio {...props} />)

      expect(wrapper.instance().value).toBe('cat')
    })

    it(`updates 'value'`, () => {
      props.defaultValue = 'cat'
      const wrapper = shallow(<Radio {...props} />)

      selectLastRadio(wrapper)

      expect(wrapper.instance().value).toBe('dog')
    })
  })
})
