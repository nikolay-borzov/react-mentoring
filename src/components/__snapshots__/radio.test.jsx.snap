// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio component if controlled renders correctly 1`] = `
<div
  className="radio-input-wrapper"
>
  <label
    className="form-label nowrap"
  >
    Radio label
  </label>
  <div
    className="radio-input radio-input--style-button"
  >
    <input
      defaultChecked={true}
      id="radioName_cat"
      name="radioName"
      onChange={[Function]}
      type="radio"
      value="cat"
    />
    <label
      className="radio-input__label"
      htmlFor="radioName_cat"
    >
      Cat
    </label>
  </div>
  <div
    className="radio-input radio-input--style-button"
  >
    <input
      defaultChecked={false}
      id="radioName_dog"
      name="radioName"
      onChange={[Function]}
      type="radio"
      value="dog"
    />
    <label
      className="radio-input__label"
      htmlFor="radioName_dog"
    >
      Dog
    </label>
  </div>
</div>
`;

exports[`Radio component if uncontrolled renders correctly 1`] = `
<div
  className="radio-input-wrapper"
>
  <label
    className="form-label nowrap"
  >
    Radio label
  </label>
  <div
    className="radio-input radio-input--style-plain"
  >
    <input
      defaultChecked={true}
      id="radioName_cat"
      name="radioName"
      onChange={[Function]}
      type="radio"
      value="cat"
    />
    <label
      className="radio-input__label"
      htmlFor="radioName_cat"
    >
      Cat
    </label>
  </div>
  <div
    className="radio-input radio-input--style-plain"
  >
    <input
      defaultChecked={false}
      id="radioName_dog"
      name="radioName"
      onChange={[Function]}
      type="radio"
      value="dog"
    />
    <label
      className="radio-input__label"
      htmlFor="radioName_dog"
    >
      Dog
    </label>
  </div>
</div>
`;
